spark3/src/main/java/org/apache/iceberg/actions/Spark3CreateAction.java

```diff
@@ -0,0 +1,187 @@
+                     CatalogPlugin destCatalog, Identifier destTableIdent) {
+
+    this.spark = spark;
+    this.sourceCatalog = checkSourceCatalog(sourceCatalog);
+    this.sourceTableIdent = sourceTableIdent;
+    this.destCatalog = checkDestinationCatalog(destCatalog);
+    this.destTableIdent = destTableIdent;
+
+    try {
+      this.sourceTable = (V1Table) this.sourceCatalog.loadTable(sourceTableIdent);
+      this.sourceCatalogTable = sourceTable.v1Table();
+    } catch (NoSuchTableException e) {
+      throw new IllegalArgumentException(String.format("Cannot not find source table %s", sourceTableIdent), e);
+    } catch (ClassCastException e) {
+      throw new IllegalArgumentException(String.format("Cannot use non-v1 table %s as a source", sourceTableIdent), e);
+    }
+    validateSourceTable();
+
+    this.sourceTableLocation = CatalogUtils.URIToString(sourceCatalogTable.storage().locationUri().get());
+  }
+
+  @Override
+  public CreateAction withProperties(Map<String, String> properties) {
+    this.additionalProperties.putAll(properties);
+    return this;
+  }
+
+  @Override
+  public CreateAction withProperty(String key, String value) {
+    this.additionalProperties.put(key, value);
+    return this;
+  }
+
+  protected SparkSession spark() {
+    return spark;
+  }
+
+  protected String sourceTableLocation() {
+    return sourceTableLocation;
+  }
+
+  protected CatalogTable v1SourceTable() {
+    return sourceCatalogTable;
+  }
+
+  protected TableCatalog sourceCatalog() {
+    return sourceCatalog;
+  }
+
+  protected Identifier sourceTableIdent() {
+    return sourceTableIdent;
+  }
+
+  protected StagingTableCatalog destCatalog() {
+    return destCatalog;
+  }
+
+  protected Identifier destTableIdent() {
+    return destTableIdent;
+  }
+
+  protected Map<String, String> additionalProperties() {
+    return additionalProperties;
+  }
+
+  private void validateSourceTable() {
+    String sourceTableProvider = sourceCatalogTable.provider().get().toLowerCase(Locale.ROOT);
+    Preconditions.checkArgument(ALLOWED_SOURCES.contains(sourceTableProvider),
+          "Cannot create an Iceberg table from source provider: %s", sourceTableProvider);
+    Preconditions.checkArgument(!sourceCatalogTable.storage().locationUri().isEmpty(),
+        "Cannot create an Iceberg table from a source without an explicit location");
+  }
+
+  private StagingTableCatalog checkDestinationCatalog(CatalogPlugin catalog) {
+    Preconditions.checkArgument(catalog instanceof SparkSessionCatalog || catalog instanceof SparkCatalog,
+        "Cannot create Iceberg table in non Iceberg Catalog. " +
+            "Catalog %s was of class %s but %s or %s are required",
+        catalog.name(), catalog.getClass(), SparkSessionCatalog.class.getName(), SparkCatalog.class.getName());
+
+    return (StagingTableCatalog) catalog;
+  }
+
+  protected StagedTable stageDestTable() {
```

Comment:
The return value of this needs to be a `SparkTable` for both uses, so I think it should return a `SparkTable` to avoid unchecked casts in the callers.

Reply:
It also needs to be StagedTable :/ since we need to commit changes (and possible abort changes)

Reply:
Can it be `StagedSparkTable` that also extends `SparkTable`?

Reply:
Didn't see we had that, yeah we should use that